### PR TITLE
TaskVine Integration

### PR DESCRIPTION
**Pull request by tphung3**
# Description

Summary: Integration of TaskVine as an executor to Parsl.
Motivation: TaskVine is the next-generation workflow execution engine from the Cooperative Computing Lab at the University of Notre Dame. This PR's goal is to ensure the interoperability between TaskVine and Parsl, as applications will soon port to TaskVine.

## Type of change

- New feature (non-breaking change that adds functionality)

### CODE REVIEW

**Comment by tphung3** [comment]
I have tried to minimize changes to current code and add tests to TaskVine. However I'm not sure how the current testing infrastructure of Parsl works. Local tests run fine but github checks are failing probably due to the reason above. Does anyone have any suggestions? Tagging relevant contributors @btovar @benclifford @kylechard. Thanks in advance!

**Comment by benclifford** [comment]
maybe you did not intend to include this subtree: 
parsl/tests/taskvine_tests/runinfo/

**Comment by tphung3** [comment]
> maybe you did not intend to include this subtree: parsl/tests/taskvine_tests/runinfo/

Yes I just removed the subtree locally and it will be gone by the next push.

**Comment by benclifford** [comment]
it's failing at the `flake8` stage which is a linter and basic semantic analysis - you can run "flake8 parsl/" on your own machine, or look click on the red cross and read the results that appear here: https://github.com/Parsl/parsl/actions/runs/4126181572/jobs/7127755700

Those flake8 errors are mostly indentation that flake8 thinks is out of compliance with PEP8, but there are a couple of other problems in there:

parsl/executors/taskvine/executor.py:50:5: F401 'taskvine.VINE_ALLOCATION_MODE_MAX_THROUGHPUT' imported but unused
parsl/executors/taskvine/executor.py:840:52: F821 undefined name 'taskvine_ALLOCATION_MODE_MAX_THROUGHPUT'

which, if that passes your local testing, suggests there's maybe some unused/untested code there?

**Comment by benclifford** [comment]
mypy type checking is failing here:

> parsl/executors/taskvine/executor.py:46: error: Cannot find implementation or library stub for module named "taskvine"  [import]

because there are no type definitions available for `taskvine`.

You can tell mypy to ignore that, which is how the python `work_queue` module is configured in parsl, by adding a block like this to the mypy configuration file:

```
[mypy-taskvine.*]
ignore_missing_imports = True
```

which looks like the existing mypy-work_queue.* block in there.

**Comment by benclifford** [comment]
`pytest` will identify which tests to run by looking for functions called `test_<something>`. That's why it's running all of those tests in taskvine_tests/

The reason this doesn't happen with workqueue is because there is a horrible bodge in parsl/tests/conftest.py called `def pytest_ignore_collect(path):`. If you want to recreate that behaviour with taskvine, I hope you can see from that function what to add. Basically it's a manual blacklist of test directories that should be ignored by pytest.

The workqueue_tests/* tests are *never* run by pytest!  (so, in practice, they are never run by anyone).

So you could add taskvine to that blacklist.

And then they would never be run by pytest, or in practice by anyone.

However:

Personally I would recommend only adding tests which are going to be run regularly by our CI system (or some other automated system, for example at cctools HQ). 

Tests which are not run are a burden on future maintenance and understanding. and I would prefer we do not add more of them; and I have been slowly removing them from other parts of the parsl test suite.

**Comment by benclifford** [comment]
The main work queue testing in our CI happens with this line in the Makefile:

```
        PYTHONPATH=.:/tmp/cctools/lib/python3.8/site-packages  pytest parsl/tests/ -k "not cleannet and not issue363" --config parsl/tests/configs/workqueue_ex.py --random-order
```

which runs a few hundred parsl tasks trying out various things.

So that's probably the main thing to concentrate on for parsl+taskvine testing.

**Comment by tphung3** [comment]
Thanks @benclifford , I just knew about the blacklisting of workqueue tests. I also think that it's probably better to have workqueue and taskvine actually tested using Parsl testing infras. Right now workqueue and parsl integration is tested daily on cctools's packaging test repo, and maybe the next/easiest step is to add another one for taskvine. 

For the long run both workqueue and taskvine executors need additional fixes to fully integrate to Parsl. I expect that taskvine be released this March, so it's probably more productive for me to come back with a hopefully better solution to this problem.

**Comment by benclifford** [comment]
If the current taskvine code is stable enough to work with parsl + the executor in this PR, I think it would be interesting to integrate it now, with testing like the current parsl CI work queue testing

**Comment by tphung3** [comment]
Taskvine is still in development mode now, so I think it'll be a hassle to set it up to test in Parsl infras. as opposed to just pulling a tarball from ccl website. I'll add some testing to the cctools packaging repo so we'll know when Taskvine and Parsl break. Thanks @benclifford again for the prompt help!

**Comment by tphung3** [comment]
Running `flake8 parsl/` on my local machine doesn't show reported errors. @benclifford is there a custom config file for flake8 in these tests?

**Comment by benclifford** [comment]
@tpghung there's a .flake8 file in the root of the repo. You might check that you are running the version that is pinned in test-requirements.txt, which is `flake8==3.8.0` - wrong flake8 and mypy versions have been the most common source of one person getting an error, and another person not, and for test/dev requirements, we pin specific versions.

But, for the actual error that's showing up, in this line:

        self.blocks = {}  # type: Dict[str, str], track Parsl blocks

I'd be inclined to use a non-comment type annotation (https://peps.python.org/pep-0526/) because we don't expect to work with Python < 3.6  - eg see this line:

https://github.com/Parsl/parsl/blob/fbdc09abc10ff5da53afe2da66879b1bdc22d1d6/parsl/dataflow/dflow.py#L201

**Comment by tphung3** [comment]
@benclifford Can you restart the tests? I'm not sure why the tests are failing. Otherwise I think this PR is ready to be reviewed.

**Comment by benclifford** [comment]
@tphung3 what does ls  /scratch365/tphung/parsl_taskvine_integration/parsl/parsl/tests show?

**Comment by tphung3** [comment]
> @tphung3 what does ls /scratch365/tphung/parsl_taskvine_integration/parsl/parsl/tests show?

```
(parsl-vine) [tphung@condorfe:parsl]-[taskvine-integration]$ ls /scratch365/tphung/parsl_taskvine_integration/parsl/parsl/tests
callables_helper.py  setup_path.sh           test_docs            test_staging
cleanup              sites                   test_error_handling  test_summary.py
configs              site_tests              test_flowcontrol     test_swift.py
conftest.py          taskvine_tests          test_flux.py         test_thread_parallelism.py
__init__.py          test_aalst_patterns.py  test_manual          test_threads
integration          test_bash_apps          test_monitoring      test-viz.sh
lint-inits.sh        test_callables.py       test_providers       utils.py
manual_tests         test_channels           test_python_apps     workqueue_tests
__pycache__          test_checkpointing      test_regression
pytest.ini           test_data               test_scaling
```

**Comment by benclifford** [comment]
Ok that's weird. In CI, tests are getting much further than that.

**Comment by benclifford** [comment]
So right now this PR is failing because the work queue tests are amended by this PR to use this test configuration, but this PR does not add that file: parsl/tests/configs/wqex_local.py

**Comment by tphung3** [comment]
A quick update: for some reason `pytest` can collect tests if `parsl` is on `/tmp`, but not if `parsl` is on afs or panfs. Have you seen any similar behavior before with `parsl` @benclifford ?

**Comment by tphung3** [comment]
Here's some relevant output in `pytestdebug.log` from running `pytest --debug parsl/tests/ -k "not cleannet" --config parsl/tests/configs/local_threads.py --random-order` (note that the `debug` option is on):

With `parsl` on `/afs`:
```
        pytest_make_collect_report [hook]
            collector: <Session tests exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=0>
        processing argument (local('/afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests'), []) [collection]
            pytest_ignore_collect [hook]
                path: /afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests/__init__.py
                config: <_pytest.config.Config object at 0x7f202b4ac670>
            finish pytest_ignore_collect --> True [hook]
            pytest_ignore_collect [hook]
                path: /afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests/__init__.py
                config: <_pytest.config.Config object at 0x7f202b4ac670>
            finish pytest_ignore_collect --> True [hook]
            pytest_ignore_collect [hook]
                path: /afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests/__init__.py
                config: <_pytest.config.Config object at 0x7f202b4ac670>
            finish pytest_ignore_collect --> True [hook]
            pytest_ignore_collect [hook]
                path: /afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests/callables_helper.py
                config: <_pytest.config.Config object at 0x7f202b4ac670>
            finish pytest_ignore_collect --> True [hook]
            pytest_ignore_collect [hook]
                path: /afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests/cleanup
                config: <_pytest.config.Config object at 0x7f202b4ac670>
            finish pytest_ignore_collect --> True [hook]
            pytest_ignore_collect [hook]
                path: /afs/crc.nd.edu/group/ccl/work/tphung/parsl_taskvine_integration/parsl/parsl/tests/conftest.py
                config: <_pytest.config.Config object at 0x7f202b4ac670>
            finish pytest_ignore_collect --> True [hook]
            ...
```
`pytest_ignore_collect` always returns True.

With `parsl` on `/tmp`:
```
        pytest_make_collect_report [hook]
            collector: <Session tests exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=0>
        processing argument (local('/tmp/parsl/parsl/tests'), []) [collection]
            pytest_ignore_collect [hook]
                path: /tmp/parsl/parsl/tests/__init__.py
                config: <_pytest.config.Config object at 0x7fa6d9c52a30>
            finish pytest_ignore_collect --> False [hook]
            pytest_collect_file [hook]
                path: /tmp/parsl/parsl/tests/__init__.py
                parent: <Session tests exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=0>
              pytest_pycollect_makemodule [hook]
                  path: /tmp/parsl/parsl/tests/__init__.py
                  parent: <Session tests exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=0>
              finish pytest_pycollect_makemodule --> <Package tests> [hook]
            finish pytest_collect_file --> [<Package tests>] [hook]
            pytest_ignore_collect [hook]
                path: /tmp/parsl/parsl/tests/__init__.py
                config: <_pytest.config.Config object at 0x7fa6d9c52a30>
            finish pytest_ignore_collect --> False [hook]
            pytest_collect_file [hook]
                path: /tmp/parsl/parsl/tests/__init__.py
                parent: <Session tests exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=0>
              pytest_pycollect_makemodule [hook]
                  path: /tmp/parsl/parsl/tests/__init__.py
                  parent: <Session tests exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=0>
              finish pytest_pycollect_makemodule --> <Package tests> [hook]
            finish pytest_collect_file --> [<Package tests>] [hook]
            pytest_ignore_collect [hook]
                path: /tmp/parsl/parsl/tests/configs
                config: <_pytest.config.Config object at 0x7fa6d9c52a30>
            finish pytest_ignore_collect --> False [hook]
          early skip of rewriting module: py._code [assertion]
          early skip of rewriting module: py._code.code [assertion]
          early skip of rewriting module: repr [assertion]
            pytest_ignore_collect [hook]
                path: /tmp/parsl/parsl/tests/configs/__init__.py
                config: <_pytest.config.Config object at 0x7fa6d9c52a30>
            finish pytest_ignore_collect --> False [hook]
            ...
```
Now `pytest_ignore_collect` always returns False.

The custom `pytest_ignore_collect` function in `parsl/parsl/tests/conftest.py` doesn't seem to have anything to do with it. Any insight is greatly appreciated. Or maybe I should just give up and avoid shared filesystems entirely...

**Comment by benclifford** [comment]
ok, i was wondering when that code would come back to bite.

the ignore hook is hardcoded to ignore some directories in the test suite. it doesn't do that very well:


```

def pytest_ignore_collect(path):
    if 'integration' in path.strpath:
        return True
    elif 'manual_tests' in path.strpath:
        return True
    elif 'workqueue_tests/test_scale' in path.strpath:
        return True
    else:
        return False
```

and it turns out anyone using the word `integration` anywhere in their directory path means each test will be ignored.

This is pretty terrible and I'll fix it, but in the meantime, change the word "integration" into a word that is not "integration", "manual_tests", or "workqueue_tests/test_scale".

**Comment by benclifford** [comment]
I opened this https://github.com/Parsl/parsl/issues/2714

**Comment by tphung3** [comment]
I'm not sure why one test was cancelled but this PR seems good to go.

**Comment by tphung3** [comment]
It seems that CI tests time out, but otherwise this PR is ready to be reviewed.

**Comment by tphung3** [comment]
@benclifford what do you think? Is it ready to be merged?

File: parsl/executors/workqueue/executor.py [hunk 1]
```diff
@@ -43,10 +43,10 @@
 from collections import namedtuple
 
 try:
-    import work_queue as wq
-    from work_queue import WorkQueue
-    from work_queue import WORK_QUEUE_DEFAULT_PORT
-    from work_queue import WORK_QUEUE_ALLOCATION_MODE_MAX_THROUGHPUT
+    import ndcctools.work_queue as wq
```
Comment (benclifford) [inline parsl/executors/workqueue/executor.py:46]: Does this change the lower bound of which version of ndcctools parsl will work with?

**Review by benclifford — CHANGES_REQUESTED** [review]

File: parsl/executors/taskvine/exec_parsl_function.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
+from parsl.app.errors import RemoteExceptionWrapper
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/exec_parsl_function.py:1]: I'm thinking about having a separate dir for ndcctools executors, like this:
`parsl/executors/ndcctools/taskvine` and `parsl/executors/ndcctools/workqueue`.
The helper scripts `exec_parsl_function.py` and `install-ndcctools.sh` will then live in `parsl/executors/ndcctools`.
But considering that workqueue is sort of going away, perhaps it's better to separate them now so less work is done in the future.
 Will Parsl continue to support WQ executor? If so then I'll do what's described above.

File: parsl/executors/taskvine/executor.py [hunk 3]
```diff
@@ -0,0 +1,1051 @@
...
+
+        full_debug: bool
+            Whether to enable full debug mode for monitoring in TaskVine.
+            Default is False.
+
+        provider: ExecutionProvider
+            The Parsl provider that will spawn worker processes.
+            Default to spawning one local vine worker process.
+
+        storage_access: Redundant argument to comply with other executors.
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/executor.py:220]: Thanks for the explanation! It makes sense to me so I fixed it.

File: parsl/executors/taskvine/executor.py [hunk 4]
```diff
@@ -0,0 +1,1051 @@
...
+        address: str
+            The ip to contact this taskvine master process.
+            If not given, uses the address of the current machine as returned
+            by socket.gethostname().
+            Ignored if project_name is specified.
+
+        port: int
+            TCP port on Parsl submission machine for TaskVine workers
+            to connect to. Workers will specify this port number when
+            trying to connect to Parsl. Default is 9123.
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/executor.py:118]: The feature is already there, so I only added more documentation for port numbers in taskvine. Fixed.

File: parsl/executors/taskvine/executor.py [hunk 5]
```diff
@@ -0,0 +1,1051 @@
...
+            Default to spawning one local vine worker process.
+
+        storage_access: Redundant argument to comply with other executors.
+            Default is None. Not used.
+    """
+
+    radio_mode = "filesystem"
+
+    @typeguard.typechecked
+    def __init__(self,
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/executor.py:227]: The coprocess model is different for TaskVine from what I've heard (WorkQueue binds coprocess with each worker deployment, while TaskVine aims to have the manager send coprocess to worker instead.) This feature probably should have its own PR.

File: parsl/executors/taskvine/executor.py [hunk 6]
```diff
@@ -0,0 +1,1051 @@
...
+        self.task_counter = -1  # task id starts from 0
+        self.should_stop = multiprocessing.Value(c_bool, False)
+
+        # mapping of function's unique memory address to its solved environment
+        self.cached_envs: Dict[int, str] = {}
+
+        if not self.address:
+            self.address = socket.gethostname()
+
+        if self.project_password_file is not None:
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/executor.py:299]: That makes sense. I fixed it.

File: parsl/executors/taskvine/executor.py [hunk 7]
```diff
@@ -0,0 +1,1051 @@
...
+            if os.path.exists(self.project_password_file) is False:
+                logger.debug("Password File does not exist, no file used")
+                self.project_password_file = None
+
+        # Build foundations of the launch command
+        self.launch_cmd = ("python3 exec_parsl_function.py {mapping} {function} {result}")
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/executor.py:305]: > It looks like the taskvine executor looks up the path for poncho_package_run, stages it, but I'm unsure if it actually gets used? Does the packages environment get launched differently in taskvine?

Yes and yes.

In WQ `poncho_package_run` is run every time a task launches to create an environment so it's expensive. TaskVine separates environment creation from task execution so executions can share environments. An environment creation is now a mini-task, of which output (in this case, the environment) is sharable between tasks.

`poncho_package_run` is declared as a TaskVine file in line 824, the mini-task (env creation) is declared if possible in line 868 (`declare_poncho`), and both of these TaskVine files are bound to the task in lines 873-875 (`add_environment` and `add_input`).

File: parsl/executors/taskvine/executor.py [hunk 8]
```diff
@@ -0,0 +1,1051 @@
...
+        self.storage_access = storage_access
+
+        # Queue to send tasks from TaskVine executor process to TaskVine manager process
+        self.task_queue: multiprocessing.Queue = multiprocessing.Queue()
+
+        # Queue to send tasks from TaskVine manager process to TaskVine executor process
+        self.collector_queue: multiprocessing.Queue = multiprocessing.Queue()
+
+        self.blocks: Dict[str, str] = {}  # track Parsl blocks
+        self.task_counter = -1  # task id starts from 0
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/taskvine/executor.py:290]: This makes sense. Changes made.

File: parsl/executors/workqueue/executor.py (hunk 1 above)
Comment (tphung3) [reply parsl/executors/workqueue/executor.py:46]: Yes it should. Where do I update the ndcctools' version? It is not in `requirements.txt`.

File: parsl/tests/configs/taskvine_blocks.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+from parsl.config import Config
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/tests/configs/taskvine_blocks.py:1]: This makes sense. `taskvine_blocks.py` deleted.

[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/tests/configs/taskvine_blocks.py:1]: Also wq_blocks.py

File: parsl/executors/workqueue/executor.py (hunk 1 above)
Comment (tphung3) [reply parsl/executors/workqueue/executor.py:46]: The lower bound is now ndcctools release 7.5.3

**Comment by tphung3** [comment]
@benclifford Do you have any comments on the updates?

**Comment by benclifford** [comment]
I think this is probably fine.

My main concern has been/is that this PR sneaks in unrelated breaking changes to the Work Queue executor (because there are production users of that executor)

**Comment by benclifford** [comment]
On this branch, my pytest runs are ending with:

```
--- Logging error ---
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/logging/__init__.py", line 1113, in emit
    stream.write(msg + self.terminator)
ValueError: I/O operation on closed file.
Call stack:
  File "/home/benc/parsl/src/parsl/parsl/dataflow/dflow.py", line 1131, in atexit_cleanup
    logger.info("python process is exiting, but DFK has already been cleaned up")
Message: 'python process is exiting, but DFK has already been cleaned up'
Arguments: ()
```

even for non-ndcctools related test runs (eg. `pytest parsl/tests/ -k "not cleannet and not issue363" --config parsl/tests/configs/htex_local.py --random-order`). I don't see this if I run against `origin/master` in an otherwise-the-same environment. I'm investigating that now.

**Comment by tphung3** [comment]
I think you're right. For now `import work_queue` only prints a warning message so I'll revert the import statements back.

**Comment by benclifford** [comment]
probably good to increase:     timeout-minutes: 30
in .github/workflows/ci.yaml to 40 minutes or so, because that 30 minute timeout is being hit a lot.

(longer term, I'd prefer both wq and task vine use coprocesses and whatever else is necessary to make them execute in the same sort of time as the htex tests, but that is not for this PR)

**Comment by benclifford** [comment]
that atexit_cleanup message happens in any test where taskvine is imported (so any test, because taskvine is always imported). getting rid of all these makes it stop happening, so I'll have a look inside those modules to see if I can understand what's happening:

```
try:
    from ndcctools.taskvine import cvine
    # from ndcctools.taskvine import Manager
    # from ndcctools.taskvine import Task
    # from ndcctools.taskvine.cvine import VINE_DEFAULT_PORT
    # from ndcctools.taskvine.cvine import VINE_ALLOCATION_MODE_MAX_THROUGHPUT
```

**Comment by benclifford** [comment]
poncho/package_create.py is doing this at line 23 at import time:

> logging.basicConfig(level=logging.INFO, format='%(asctime)s:%(levelname)s:%(message)s')

I would prefer that parsl (and by extension, anything that is imported into parsl) not modify the log configuration of the environment it is imported into.

This comes, I think from ndcc.taskvine importing poncho package_serverize which imports package_create...

**Comment by benclifford** [comment]
one workaround for that poncho log config would be to not include the taskvine executor in the default parsl.*, parsl.executors.* import list - that way, its imports (and side-effects) would only happen if a user did `import parsl.executors.taskvine`, I think.

**Comment by tphung3** [comment]
Update: I just figured out how to stop `poncho` from hijacking logging config so the above problem now goes away (see https://github.com/cooperative-computing-lab/cctools/pull/3375). This change won't reflect until the next release however. 

Isolating TaskVine works with Parsl tests but has consequences (like reverting test cases). I'll meet Dr. Thain tomorrow to discuss the available actions...

**Comment by benclifford** [comment]
what problems do you get if you remove taskvine from parsl/executors/__init__.py and parsl/__init__.py? I would hope that that would work with the existing test infrastructure (with changed imports)

**Comment by tphung3** [comment]
flake8 and mypy jump in for static analysis and find out that TaskVineExecutor is not there anymore. I can design a fake TaskVineExecutor class to fool it but it will be ugly in the main source code.

**Comment by benclifford** [comment]
I was imagining leaving all the code in the place it is now, but removing the import statements that make it imported by default - so to use taskvine, you need a longer import statement rather than importing from parsl.executors top level.

I'll make a branch testing out what I'm thinking and paste it here if it seems to work.

**Comment by benclifford** [comment]
@tphung3 this commit is what I was imagining for the current fix:

 https://github.com/Parsl/parsl/commit/9af0ebbda0ae592048b870f3a928137f61207697

Users who do not touch taskvine do not get taskvine imported when they `import parsl`, so the logging problem does not happen.

Users have to explicitly say `from parsl.executors.taskvine import TaskVine` - as I have changed in the test cases - and in that case, the logging gets messed up. but only if they are opting into taskvine.

**Comment by tphung3** [comment]
I think this PR is ready to merge.

**Pull request merged**